client.py

```diff
@@ -100,37 +129,47 @@ def close(self):
 
     # **** API end ****
 
-    # qos==2 Retransmit until matching ACK received
-    async def _do_qos(self, mid, line):
+    async def _write(self, preheader, header, buf, qos, mid, init=False):
+        if buf is None:
+            buf = b""
         while True:
-            while not self._ok:  # Wait for any outage to clear
+            while not self._ok and init is False:
```

Comment:
Messages are never corrupted. If the client_id is lost it will behave as per an initial server outage.

Reply:
It's not about message corruption, it's about making sure that the server is actually responding correctly and getting the message. If he does not get the client_id because it somehow got lost during the transmission, it will accept the fist message as client_id which could screw up your application. This is of course a more paranoid approach but I don't see a point in having qos messages if the client identifying message does not even use qos.

Reply:
That's fine, but my version resolves this as follows. The client waits until it gets a keepalive from the server before it sends any data. The server won't send a KA until after it gets the client ID. This is how the client detects the "server up" state.

Reply:
You are right, I did not think about the keepalive as an ACK to the client_id. That should indeed be enough and make the server easier (and the client a little bit).
The only problem I'm worried about is that the client keeps waiting for the keepalive instead of closing the connection if it does not receive one after a timeout. So if the client_id gets lost, it will hang forever.

Reply:
In my version **self._ok** is **False** until a packet is received from the server.

The continuously running **._reader** calls **.readline** in a loop.  **._readline** will throw an **OSError** if it times out without receiving anything back. **.reader** will trap this and set **._evfail**.

So if the client ID is lost, client behaviour is exactly the same as if the server was down. It times out and goes through the fail process of shutting everything down before re-trying.

Reply:
Of course, I'm sorry. Should have known that.